ADDED: Draw Distance toggle under Experimental enhancement + Bugfix

PurpleHato opened the pull request:
ADDED; Button under Enhancement --> Experimental
ADDED: Kokiri Draw Distance specific button for purists (since it is intentional that they spawn when you have your face on them)
FIXED: Dark Link will now be able to spawn during his fight on the Water Temple

PurpleHato committed:
ADDED: Draw Distance toggle under Experimental enhancement

PurpleHato committed:
FIXED: Dark Link Apparition + Kokiri NPC spawn

MelonSpeedruns commented:
Legitimate!

PurpleHato committed:
Added a Toggle for Kokiri NPC since they are mystics

PurpleHato committed:
TWKEA: Desciription got the Kokiri draw distance

ghost commented:
Nice work! When I compiled this Epona now makes a neigh sound at the start of Title Screen. Can you please check this is not due to this feature? Thanks

PurpleHato commented:
> Nice work! When I compiled this Epona now makes a neigh sound at the start of Title Screen. Can you please check this is not due to this feature? Thanks

It is due to that since you will "see" the Epona spawning which is always followed by her shout, same thing will happen for the bushes Epona will encounter while following her path you'll hear them being destroyed

Kinda lazy to fix that specific part to be honest, will probably check that later

leggettc18 commented:
Speaking of Epona, I think this might be messing with using Epona's song to summon her when she's in the same scene as you. Normally with the draw distance when you play Epona's song, she isn't loaded in yet so she kind of teleports to you and then runs toward you, or she's close enough that her running from where she is spawned in isn't very far away. But with draw distance turned off, she's already loaded in even if she's all the way across Hyrule Field, so she does hear the song and run toward you, but she only runs so far. When I played the song I thought it was glitched until I saw her way off in the background and realized it was probably the draw distance that caused it.

ADDENDUM: I actually just dropped Epona off in Gerudo Valley specifically to see if her not being already on Hyrule Field caused her to spawn properly, and she still did not spawn in, at least not anywhere where I could see her. Turning the draw distance back on immediately fixed it, and the next Epona's song spawned her in correctly.

Another observation, not necessarily worth the energy to fix, when you wake up Biggoron during the quest to get the Prescription, you can hear him groaning and rubbing his eyes the entire way down the mountain. I suppose that's somewhat more realistic given his size, just thought it was funny.

jbodner09 commented on libultraship/libultraship/SohImGuiImpl.cpp:
```diff
@@ -762,6 +762,12 @@ namespace SohImGui {
                 EnhancementCheckbox("Disable LOD", "gDisableLOD");
                 Tooltip("Turns off the level of detail setting, making models always use their higher poly variants");
+                EnhancementCheckbox("Disable Draw Distance", "gDisableDrawDistance");
+                Tooltip("Turns off the objects draw distance, making objects being visible from a longer range");
```
Language tweak:  "Turns off the draw distance limitation, making objects visible from further away"

jbodner09 commented on libultraship/libultraship/SohImGuiImpl.cpp:
```diff
@@ -762,6 +762,12 @@ namespace SohImGui {
+                Tooltip("Turns off the objects draw distance, making objects being visible from a longer range");
+                if (CVar_GetS32("gDisableDrawDistance", 0) == 1) {
+                    EnhancementCheckbox("Kokiri Draw Distance", "gDisableKokiriDrawDistance");
+                    Tooltip("Kokiris are mystical being that appear from a certain distance\nEnable this will remove their draw distance\nNeeds to reload the map to take effect");
```
Language tweak:  "Kokiri are mystical beings that only appear from a certain distance away
Enabling this will remove their draw distance limitation
Requires reloading the map to take effect"

PurpleHato committed:
FIXED: Horses behavior

Fixed Epona shout on title screen + Epona call from songFixed Zelda and Ganon horses that could still be seen on camera pan on the castle escape cinematic

leggettc18 commented:
Can confirm that last commit fixes Epona's issues.

PurpleHato commented:
> Can confirm that last commit fixes Epona's issues.

Since you seems to play test it

Please let me know if there is any behaviours that needs to be fixed like Biggoron

Same if you see some models not despawning on cinematic pans when they should have 😃

leggettc18 commented:
Sure, I started using the build from Jenkins for this in a semi-casual playthrough recently, switched to this around the time I went to Adult Link for the first time. I'll keep using it and let you know what all I find. I've beaten Water Temple and Forest Temple so far (doing dungeons out of order), and done Biggoron's quest (some of that may have been done before the switch. Then once this file is beaten I'll at least check some key points as Young Link.

A common thing I've noticed so far is that you can often hear noisy actors in large areas that you wouldn't hear in Vanilla since they normally stop getting drawn altogether when you get too far away. I.E. Biggoron when he's rubbing his eyes or the Blue Bubbles at the top of the block pushing room in Forest Temple. It doesn't seem to have any other noticeable effects on gameplay, but sometimes the noise is a bit annoying. Would it be feasible to repurpose the Draw Distance into an Earshot Distance in a generic enough way that it can apply to all actors? Or is it easier to handle these on a case by case basis? Or is this just a necessary evil to have unlimited draw distance?

PurpleHato commented:
> Sure, I started using the build from Jenkins for this in a semi-casual playthrough recently, switched to this around the time I went to Adult Link for the first time. I'll keep using it and let you know what all I find. I've beaten Water Temple and Forest Temple so far (doing dungeons out of order), and done Biggoron's quest (some of that may have been done before the switch. Then once this file is beaten I'll at least check some key points as Young Link.
> 
> A common thing I've noticed so far is that you can often hear noisy actors in large areas that you wouldn't hear in Vanilla since they normally stop getting drawn altogether when you get too far away. I.E. Biggoron when he's rubbing his eyes or the Blue Bubbles at the top of the block pushing room in Forest Temple. It doesn't seem to have any other noticeable effects on gameplay, but sometimes the noise is a bit annoying. Would it be feasible to repurpose the Draw Distance into an Earshot Distance in a generic enough way that it can apply to all actors? Or is it easier to handle these on a case by case basis? Or is this just a necessary evil to have unlimited draw distance?

Case by case is probably the way to do it and simply remove the noisy  ones out of the draw distance removal is the easier way since their sounds are basically from the moment you see them and not distance based, I'll need to analyze that to see if it's possible to do like a "range" from where the sound should play for problematics entities

MelonSpeedruns reviewed (approved):
LGTM

leggettc18 commented:
Finished my playthrough, didn't notice anything else out of the ordinary with this enhancement, Tomorrow I will also go through some of the child areas that I passed before this PR came out but I can't think of anything that would be an issue. I did already experience most of the dungeons again while searching for a few Heart Pieces and Gold Skulltulas I missed and there weren't any issues.

As for the audio issues we found earlier, I'm sure excepting them from the draw distance removal is probably the easiest, but don't Gold Skulltulas already have a system in place for proximity based audio? Might be possible to adapt that code, but whether or not its worth the effort is your decision.

ghost commented:
Can you add no draw distance to King Zora? King Zora spawns as you're walking up the stairs. Thanks

Also, the Deku Tree isn't always drawn.

Is this even possible? as it looks to be based on collision trigger?

PurpleHato commented:
> Can you add no draw distance to King Zora? King Zora spawns as you're walking up the stairs. Thanks
> 
> Also, the Deku Tree isn't always drawn.
> 
> Is this even possible? as it looks to be based on collision trigger?

This is a complete different process
Each scenes (map) are divided by multiple rooms, that load / unload from a trigger point
Example in a dungeon, when you open the door, you're unloading your current room to load the one you're going in

What you want is a complete rewrite of the scene loading system and not a draw visibility distance

Perhaps someone will do that in the future which will make this feature compatible with it if it ever happen

PurpleHato committed:
Merge branch 'HarbourMasters:develop' into DrawDistance

PurpleHato committed:
Merge branch 'HarbourMasters:develop' into DrawDistance

Marcelo20XX commented:
I always do my playthroughs using the Draw Distance cvar on and I didn't see anything out of the ordinary save for Biggoron or the Ganon and Zelda horses not disappearing in their cinematic...

But one thing I noticed that would be nice to add now that Draw Distance is added to the enhancements section would be a way to tweak the spawning distance of more hardcoded actors, I am speaking for example about Mido and Saria, they aren't spawning as the rest of the Kokiri because the function is not there, also some grass patches, mostly in Hyrule Field, are still spawning only when Link gets a certain distance...
